### ISSUE TITLE
false positive of SC1035 for `!!`

### ISSUE THREAD

**Issue by despairblue** [koalaman/shellcheck#658]
```shell
while [ $? -ne 0 ]; do !!; echo retry; sleep 5; done
```

This will retry the last command every 5 seconds until it succeeds. Adding a whitespace between the two exclamation marks does alter the meaning of this, to quote the rationale behind the rule:

> Bourne shells are very whitespace sensitive. Adding or removing spaces can drastically alter the meaning of a script. In these cases, ShellCheck has noticed that you're missing a space at the position indicated.

Thanks for making this, btw :+1:

**Comment by kurahaupo** [comment]
> ```shell
> while [ $? -ne 0 ]; do !!; echo retry; sleep 5; done
> ```
>
> This will retry the last command every 5 seconds until it succeeds.

In an interactive shell, that will test the previous command, and if it
failed, will repeat it indefinitely without regard for whether it succeeds
or fails, until the sleep gets interrupted.

That is not a scripting fragment as such, because history expansion is only
done in interactive shells by default, and even if you enable it inside a
script, it would read from the history created by an interactive shell
outside the script.

Comparing ```$?``` against zero should never be necessary, and generally
indicates poorly thought out program flow.

Instead just do this:

```repeat() { while ! "$@" ; do sleep 5 ; done ; }
repeat try_some_command_here "and don't" forget the\ args
```

**Comment by koalaman** [comment]
You're right, of course. ShellCheck doesn't understand history expansion. It's not really a feature used in scripts where it's disabled by default, but I can definitely see this coming up when checking your `.bashrc`.

Supporting history expansion would probably squash more true than false positives though: `!true` is valid history expansion but probably not what the user wanted. Please ignore the warning.

**Comment by despairblue** [comment]
@kurahaupo you're right. Thanks for showing me a better way to achieve this, or rather showing me a way at all since mine was completely flawed :smile: 

@koalaman you're probably right. Though I wasn't really thinking of removing the message, but rather making `!!` an exception to the rule. I don't know how much complexity that would add since I haven't looked at the code base. That said, I'm fine with closing this since it's pretty useless in a script.

Thank you!
